numba/tests/test_dispatcher.py

```diff
@@ -934,6 +934,7 @@ def setUp(self):
     def tearDown(self):
         sys.modules.pop(self.modname, None)
         sys.path.remove(self.tempdir)
+        shutil.rmtree(self.tempdir)
```

Comment:
I think that because this class is a base class of other test classes it is now not safe for concurrent execution. The `temp_directory` call in the `setUp` is safe for concurrent use by virtue of ignoring `error.EEXIST` on `os.mkdir` calls. However, if tests were interleaved such that a `tearDown` occurred whilst another test that relied on the cache root directory were running, then the cache root directory would be unconditionally removed, this potentially causing problems.

Three potential fixes:
1. Add `_numba_parallel_test_ = False` as a member variable to this base class, this will serialize execution of all such tests (alternatively inherit from `numba.tests.support.SerialMixin` which has the same effect). This is probably less desirable as a fix as there's already a lot of execution time wasted in tests that cannot run concurrently.
2. Revoke these changes and make just the necessary changes to do removal of the directory added in the original patches and defer longer term fixes to https://github.com/numba/numba/issues/4274.
3. Something else?! Suggestions welcomed.

Reply:
Thanks for pointing that out @stuartarchibald. I prefer option 2 for this PR. As you mentioned, we can revisit any issues surrounding `temp_directory` following discussion in #4274